
SCE2 <- read10xCounts("./10Xdata/5k_pbmc_protein_v3_filtered_feature_bc_matrix.h5")

sum(rowData(SCE2)$ID != rowData(SCE)$ID) #make sure we good
if (TRUE) {
  GEX <- SCE2[rowData(SCE2)$Type == "Gene Expression"]
  counts(GEX) <- as.matrix(counts(GEX))
  GEX <- scater::normalize(GEX)
  GEX <- GEX[keep,]
  GEX <- GEX[hvg,]
  GEXm <- GEX %>%
    logcounts() %>%
    t()
  predictions <- predict(model, GEXm)
  
  realCITE <- SCE2[proteinList]
}



## READ IN / PREPARE ##########

# we have a list of predictions that iterates through each cell, then each protein
# this matrix will have proteins on the rows and cells on the columns
if (FALSE) {
  predictions <- predict(model, trainingData)
  
  #realCITE <- SCE[proteinList,(trainingSize+1):(trainingSize+testingSize)]
  realCITE <- SCE[proteinList,1:trainingSize]
  
}

predictionMatrix <- t(matrix(unlist(predictions), ncol = length(predictions)))
predCITE <- SingleCellExperiment(assays = list(logcounts = predictionMatrix))
# grab the metadata from the other
rowData(predCITE) <- rowData(SCE[proteinList,])
colData(predCITE) <- colData(SCE)
row.names(predCITE) <- row.names(SCE[proteinList,])

### DO THE THING ####################

library(scran)



GEX <- runPCA(GEX, ncomponents = 10)
GEX <- runTSNE(GEX, use_dimred = 'PCA')
GEX <- runUMAP(GEX, use_dimred = 'PCA')

g3 <- buildSNNGraph(GEX, k = 50, use.dimred = 'PCA')
clust3 <- igraph::cluster_louvain(g3)$membership
table(clust3)
GEX$cluster = as.factor(clust3)
GEX$cluster2 = as.factor(clust2)

plotReducedDim(GEX, "TSNE", colour_by="cluster") + ggtitle("TSNE of real GEX")
plotReducedDim(GEX, "UMAP", colour_by="cluster") + ggtitle("UMAP of real GEX")
plotReducedDim(GEX, "TSNE", colour_by="cluster2") + ggtitle("TSNE of real GEX colored by protein clusters")

counts(realCITE) <- as.matrix(counts(realCITE))


logcounts(realCITE) <- clr(counts(realCITE))
realCITE <- runPCA(realCITE, ncomponents = 10)
realCITE = runTSNE(realCITE, use_dimred = 'PCA')
realCITE <-  runUMAP(realCITE, use_dimred = 'PCA')

g2 <- buildSNNGraph(realCITE, k = 50, use.dimred = 'PCA')
clust2 <- igraph::cluster_louvain(g2)$membership
table(clust2)
realCITE$cluster = as.factor(clust2)
realCITE$cluster3 = as.factor(clust3)

plotReducedDim(realCITE, "TSNE", colour_by="cluster") + ggtitle("TSNE of real protein")
plotReducedDim(realCITE, "TSNE", colour_by="cluster3") + ggtitle("TSNE of real protein colored by GEX clusters")






predCITE <- runPCA(predCITE, ncomponents = 10)
predCITE <- runTSNE(predCITE, use_dimred = 'PCA')
predCITE <- runUMAP(predCITE, use_dimred = 'PCA')

g1 <- buildSNNGraph(predCITE, k = 50, use.dimred = 'PCA')
clust1 <- igraph::cluster_louvain(g1)$membership
table(clust1)
predCITE$cluster = as.factor(clust1)
predCITE$cluster2 = as.factor(clust2)
predCITE$cluster3 <- as.factor(clust3)

#
#I am an intern in the Gottardo lab, being mentored by Dr. Rob Amezquita. Our dry lab leverages statistics to analyze the big data generated by various wet labs. We focus mainly on Single Cell techniques such as Flow Cytometry and Single Cell RNA-sequencing (scRNA-seq). My main project uses machine learning to predict cell-surface protein expression values from gene expression assays (such as scRNA-seq). I am using R and running a TensorFlow neural network through Keras. 

plotReducedDim(predCITE, "TSNE", colour_by="cluster") + ggtitle("TSNE of predicted protein")
plotReducedDim(predCITE, "UMAP", colour_by="cluster") + ggtitle("UMAP of predicted protein")
plotReducedDim(predCITE, "TSNE", colour_by="cluster2") + ggtitle("TSNE of predicted protein colored by real CITE clusters")
plotReducedDim(predCITE, "UMAP", colour_by="cluster2") + ggtitle("UMAP of predicted protein colored by real CITE clusters")
plotReducedDim(predCITE, "TSNE", colour_by="cluster3") + ggtitle("TSNE of predicted protein colored by real GEX clusters")
plotReducedDim(predCITE, "UMAP", colour_by="cluster3") + ggtitle("UMAP of predicted protein colored by real GEX clusters")

plotReducedDim(predCITE, "TSNE", colour_by="CD3") + ggtitle("TSNE of predicted protein colored by CD45RA")

pplot <- function(x) {
  plotReducedDim(realCITE[,GEX$cluster == 3], "TSNE", colour_by=row.names(realCITE)[x]) + ggtitle(paste("TSNE of real protein colored by", row.names(realCITE)[x]))
}
library(gridExtra)
grid.arrange(grobs = lapply(1:9, pplot), nrow = 3)
grid.arrange(grobs = lapply(10:18, pplot), nrow = 3)
grid.arrange(grobs = lapply(19:27, pplot), nrow = 3)




