<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Deep Learning 2 • r10xDL</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/all.min.css" integrity="sha256-PbSmjxuVAzJ6FPvNYsrXygfGhNJYyZ2GktDbkMBqQZg=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/v4-shims.min.css" integrity="sha256-A6jcAdwFD48VMjlI3GDxUd+eCQa7/KWy6G9oe/ovaPA=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Deep Learning 2">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">r10xDL</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/DeepLearningWriteup.html">Deep Learning 2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Deep Learning 2</h1>
            
      
      
      <div class="hidden name"><code>DeepLearningWriteup.Rmd</code></div>

    </div>

    
    
<p>Here we will be doing a quick foray into deep learning using the 10X genomics data published in July 2019. This data tracks around 5000 PBMC cells. For each cell, the data specify gene expression data, protein expression data, and antigen specificity data.</p>
<p>Broadly, our goal is to create a machine learning approach to predict protein expression from gene expression data. This model, if it is effective, could be useful for predicting cell type, imputing sparse protein counts, or transferring to different scRNA-seq data sets to generate new protein information.</p>
<p>We have a lot of data here to work with, but it is not obvious how best to approach building the model. I decided to use a basic neural net, as they have shown excellent performance recently in transcriptomic problems, and are ideal for predicting complex nonlinear functions such as protein values. Other options include decision forests and linear regression, both of which might work here and have relative benefits such as simplicity and readability.</p>
<p>Please note that several hyperparameters have been adjusted to help the vignette run quickly. I will mark these spots with more optimal values.</p>
<p>I will be using BioConductor/SingleCellExperiment to store and process the data, before feeding it into the Keras interface to TensorFlow to build the actual neural network. I will mostly show the best setup I have found, with notes explaining how I tested and made decisions. With that said, let’s dive into the problem.</p>
<div id="cut-up-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#cut-up-the-data" class="anchor"></a>Cut up the data</h2>
<p>We will begin with processing and subsetting of the data, to get it into a form that can be fed into Keras.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">31415</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(DropletUtils)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(BiocManager)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(SingleCellExperiment)</a>
<a class="sourceLine" id="cb1-5" data-line-number="5">trainingSCE &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html">read10xCounts</a></span>(<span class="st">"~/Documents/10Xdata/5k_pbmc_protein_v3_nextgem_filtered_feature_bc_matrix.h5"</span>)</a>
<a class="sourceLine" id="cb1-6" data-line-number="6">testingSCE &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html">read10xCounts</a></span>(<span class="st">"~/Documents/10Xdata/5k_pbmc_protein_v3_filtered_feature_bc_matrix.h5"</span>)</a>
<a class="sourceLine" id="cb1-7" data-line-number="7"></a>
<a class="sourceLine" id="cb1-8" data-line-number="8">trainingSize &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(trainingSCE)</a>
<a class="sourceLine" id="cb1-9" data-line-number="9">testingSize &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(testingSCE)</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="co"># singler = CreateSinglerSeuratObject(counts =</span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="co"># 'GSE74923_series_matrix.txt', annot = 'GSE74923_types.txt',</span></a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="co"># project.name = 'GSE74923', min.genes = 500, min.cells = 2,</span></a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="co"># npca = 10, regress.out ='nUMI', technology= 'C1', species =</span></a>
<a class="sourceLine" id="cb1-15" data-line-number="15"><span class="co"># 'Mouse', citation = 'Kimmerling et al.</span></a>
<a class="sourceLine" id="cb1-16" data-line-number="16"><span class="co"># 2016',reduce.file.size = T, variable.genes = 'de',</span></a>
<a class="sourceLine" id="cb1-17" data-line-number="17"><span class="co"># normalize.gene.length = T)</span></a>
<a class="sourceLine" id="cb1-18" data-line-number="18"></a>
<a class="sourceLine" id="cb1-19" data-line-number="19"></a>
<a class="sourceLine" id="cb1-20" data-line-number="20"><span class="co"># sum(rowData(trainingSCE) != rowData(testingSCE)) This is a</span></a>
<a class="sourceLine" id="cb1-21" data-line-number="21"><span class="co"># good sanity check -&gt; the rows in each SCE are equivalent.</span></a>
<a class="sourceLine" id="cb1-22" data-line-number="22"><span class="co"># This will not always be the case</span></a>
<a class="sourceLine" id="cb1-23" data-line-number="23"></a>
<a class="sourceLine" id="cb1-24" data-line-number="24"><span class="co"># this is a boolean list with TRUE in the indexes of proteins</span></a>
<a class="sourceLine" id="cb1-25" data-line-number="25">proteinList &lt;-<span class="st"> </span>(<span class="kw">rowData</span>(trainingSCE)<span class="op">$</span>Type <span class="op">==</span><span class="st"> "Antibody Capture"</span>) <span class="op">&amp;</span><span class="st"> </span></a>
<a class="sourceLine" id="cb1-26" data-line-number="26"><span class="st">    </span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span>(<span class="st">"IgG*"</span>, <span class="kw">rowData</span>(trainingSCE)<span class="op">$</span>Symbol)</a>
<a class="sourceLine" id="cb1-27" data-line-number="27">getProteinMatrix &lt;-<span class="st"> </span><span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb1-28" data-line-number="28">    <span class="co"># from an SCE, gets the transposed matrix of proteins</span></a>
<a class="sourceLine" id="cb1-29" data-line-number="29">    x[proteinList] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-assays.html">counts</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/compositions/man/clr.html">clr</a></span>()  <span class="co"># we perform a centered log ratio transform to normalize the counts</span></a>
<a class="sourceLine" id="cb1-30" data-line-number="30">}</a>
<a class="sourceLine" id="cb1-31" data-line-number="31"></a>
<a class="sourceLine" id="cb1-32" data-line-number="32"><span class="co"># tensorflow breaks if fed dashes, so we replace dashes in</span></a>
<a class="sourceLine" id="cb1-33" data-line-number="33"><span class="co"># the protein names with underscores</span></a>
<a class="sourceLine" id="cb1-34" data-line-number="34">fixProteinColNames &lt;-<span class="st"> </span><span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb1-35" data-line-number="35">    <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(x) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">"-"</span>, <span class="st">"_"</span>, <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(x))</a>
<a class="sourceLine" id="cb1-36" data-line-number="36">    x</a>
<a class="sourceLine" id="cb1-37" data-line-number="37">}</a>
<a class="sourceLine" id="cb1-38" data-line-number="38"></a>
<a class="sourceLine" id="cb1-39" data-line-number="39"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(purrr)</a>
<a class="sourceLine" id="cb1-40" data-line-number="40"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(compositions)</a>
<a class="sourceLine" id="cb1-41" data-line-number="41">proteinTrainingData &lt;-<span class="st"> </span>trainingSCE <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">getProteinMatrix</span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb1-42" data-line-number="42"><span class="st">    </span><span class="kw">fixProteinColNames</span>()</a>
<a class="sourceLine" id="cb1-43" data-line-number="43">proteinTestingData &lt;-<span class="st"> </span>testingSCE <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">getProteinMatrix</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">fixProteinColNames</span>()</a>
<a class="sourceLine" id="cb1-44" data-line-number="44"></a>
<a class="sourceLine" id="cb1-45" data-line-number="45"><span class="kw"><a href="https://rdrr.io/r/base/rm.html">remove</a></span>(fixProteinColNames)</a>
<a class="sourceLine" id="cb1-46" data-line-number="46"></a>
<a class="sourceLine" id="cb1-47" data-line-number="47">geneTrainingSCE &lt;-<span class="st"> </span>trainingSCE[<span class="kw">rowData</span>(trainingSCE)<span class="op">$</span>Type <span class="op">==</span><span class="st"> "Gene Expression"</span>]</a>
<a class="sourceLine" id="cb1-48" data-line-number="48">geneTestingSCE &lt;-<span class="st"> </span>testingSCE[<span class="kw">rowData</span>(testingSCE)<span class="op">$</span>Type <span class="op">==</span><span class="st"> "Gene Expression"</span>]</a></code></pre></div>
</div>
<div id="cleaning-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#cleaning-the-data" class="anchor"></a>Cleaning the data</h2>
<p>Now we will clean and prepare the data to be used in the model. We use some basic techniques like filtering out poorly-expressed genes and normalizing the GEX matrix.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># just like for the proteins, we need to turn the delayed</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="co"># matrices to matrices</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-assays.html">counts</a></span>(geneTrainingSCE) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-assays.html">counts</a></span>(geneTrainingSCE))</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-assays.html">counts</a></span>(geneTestingSCE) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-assays.html">counts</a></span>(geneTestingSCE))</a>
<a class="sourceLine" id="cb2-5" data-line-number="5"></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co"># keep genes with at least 1 read in at least 1% of cells</span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co"># (remove low frequency genes)</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8">min_reads &lt;-<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9">min_cells &lt;-<span class="st"> </span><span class="fl">0.01</span> <span class="op">*</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(geneTrainingSCE)</a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="co"># calculate a boolean vector for which genes to keep</span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11">keep &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-assays.html">counts</a></span>(geneTrainingSCE) <span class="op">&gt;=</span><span class="st"> </span>min_reads) <span class="op">&gt;=</span><span class="st"> </span>min_cells</a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="co"># subset both SCEs using this vector</span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13">geneTrainingSCE &lt;-<span class="st"> </span>geneTrainingSCE[keep, ]</a>
<a class="sourceLine" id="cb2-14" data-line-number="14">geneTestingSCE &lt;-<span class="st"> </span>geneTestingSCE[keep, ]</a>
<a class="sourceLine" id="cb2-15" data-line-number="15"></a>
<a class="sourceLine" id="cb2-16" data-line-number="16"><span class="co"># for readability, swap rownames from row ID to row symbol</span></a>
<a class="sourceLine" id="cb2-17" data-line-number="17"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(scater)</a>
<a class="sourceLine" id="cb2-18" data-line-number="18">featureNames &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/scater/man/uniquifyFeatureNames.html">uniquifyFeatureNames</a></span>(<span class="dt">ID =</span> <span class="kw">rowData</span>(geneTrainingSCE)<span class="op">$</span>ID, </a>
<a class="sourceLine" id="cb2-19" data-line-number="19">    <span class="dt">names =</span> <span class="kw">rowData</span>(geneTrainingSCE)<span class="op">$</span>Symbol)</a>
<a class="sourceLine" id="cb2-20" data-line-number="20"><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(geneTrainingSCE) &lt;-<span class="st"> </span>featureNames</a>
<a class="sourceLine" id="cb2-21" data-line-number="21"><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(geneTestingSCE) &lt;-<span class="st"> </span>featureNames</a>
<a class="sourceLine" id="cb2-22" data-line-number="22"></a>
<a class="sourceLine" id="cb2-23" data-line-number="23"><span class="co"># normalize</span></a>
<a class="sourceLine" id="cb2-24" data-line-number="24">geneTrainingSCE &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scater/man/normalize.html">normalize</a></span>(geneTrainingSCE)</a>
<a class="sourceLine" id="cb2-25" data-line-number="25">geneTestingSCE &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/scater/man/normalize.html">normalize</a></span>(geneTestingSCE)</a></code></pre></div>
</div>
<div id="select-data" class="section level2">
<h2 class="hasAnchor">
<a href="#select-data" class="anchor"></a>Select Data</h2>
<p>Here is where we decide which genes and proteins to train on. I decided to train the model on only the more highly variable genes. Hopefully this reduces noise in the model and makes it easier and faster for the model to train. I also decided to remove CD127 from the list of proteins to predict, as the actual values for it are mostly 0.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># find highly variable genes</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(scran)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">fit &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/scran/man/trendVar.html">trendVar</a></span>(geneTrainingSCE, <span class="dt">use.spikes =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">dec &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/scran/man/decomposeVar.html">decomposeVar</a></span>(geneTrainingSCE, fit)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">hvg &lt;-<span class="st"> </span>dec<span class="op">$</span>bio <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>  <span class="co"># get biologically relevant genes</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"></a>
<a class="sourceLine" id="cb3-7" data-line-number="7">getGeneData &lt;-<span class="st"> </span><span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb3-8" data-line-number="8">    <span class="co"># subset both SCE's with that set of genes</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9">    x[hvg, ] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SCE-assays.html">logcounts</a></span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>()</a>
<a class="sourceLine" id="cb3-10" data-line-number="10">}</a>
<a class="sourceLine" id="cb3-11" data-line-number="11"></a>
<a class="sourceLine" id="cb3-12" data-line-number="12">geneTrainingData &lt;-<span class="st"> </span><span class="kw">getGeneData</span>(geneTrainingSCE)</a>
<a class="sourceLine" id="cb3-13" data-line-number="13">geneTestingData &lt;-<span class="st"> </span><span class="kw">getGeneData</span>(geneTestingSCE)</a>
<a class="sourceLine" id="cb3-14" data-line-number="14"></a>
<a class="sourceLine" id="cb3-15" data-line-number="15"><span class="kw"><a href="https://rdrr.io/r/base/rm.html">remove</a></span>(getGeneData)</a></code></pre></div>
</div>
<div id="building-the-model" class="section level2">
<h2 class="hasAnchor">
<a href="#building-the-model" class="anchor"></a>Building the model</h2>
<p>I tried many iterations of the model’s hyperparameters, and found some with decent performance on a few important proteins. Here is a compilation of what I know about these hyperparameters, and what they do. This list is sorted with most important first.</p>
<ul>
<li>Dropout - Dropout is the process of choosing randomly some proportion of nodes from a layer (often the input layer) to get dropped completely. This helps ensure that the model is training using all available data, as sometimes key genes are not included. Using dropout does seem to have a large positive effect on performance. Good values seem to range from ~.2 to ~.5.</li>
<li>Learning rate - The rate at which the model changes its parameters. Values between 0.0002 and 0.001 seem to work fine.
<ul>
<li>If this is too high, the model will jump between local minima and potentially never converge. In the loss graph, this correlates to great variance in loss between adjacent epochs. If the model has not converged by the end of the run, it is probably losing a lot of prediction accuracy, so this is a problem</li>
<li>If this is too low, the model will simply take a long time to converge, which is not a big issue.</li>
</ul>
</li>
<li>Layer sizes - This is one I don’t fully understand. One thing I have noticed is that too few layers in a bottleneck (less than 4?) is a bad thing. The number of nodes in a bottleneck layer represents the dimensionality through which the input space is forced. Too few, and the information carried by the bottleneck will not be enough to reconstruct important features.</li>
<li>Number of layers - Again, I haven’t seen much difference. I also haven’t spent too much time looking. Usually ~3 is a solid number. Too many less and there is no space to approximate hard non-linear functions. Too many more and the model will have a hard time making use of the added complexity.</li>
<li>Activation function - RELU is an activation function which zeros out any activation below 0, and sigmoid squishes any value asymptotically between 0 and 1. Basically the difference is that RELU destroys negative values, and sigmoid doesn’t. I dabbled briefly with sigmoid activation and it ruined everything. I have been sticking to RELU. I think it does a good job of zeroing out noise in a lot of cases. I wouldn’t be opposed to having a sigmoid layer here or there, but I do think the model requires 1 relu layer at a minimum.</li>
<li>Batch size - This is the number of samples that the model goes through before it decides which direction to train. Generally a power of 2, often 32, 64, or 128. Small sizes will mean the model is often pulled different directions towards different local minima, whereas large sizes will cause the model to move in the same direction. I find that 128 works nicely, but haven’t played with it too much.</li>
<li>Optimizer function / Loss function - I didn’t test either of these, but I will include a note about them here. The optimizer function used is RMSprop</li>
</ul>
<div class="figure">
<img src="DNNarchitecture.png" alt="A simplified view of the neural net architecture. Layer sizes from left to right were actually 64, 32, 4, 8, 4, 1, as opposed to the 4, 3, 2, 3, 2, 1 shown here."><p class="caption">A simplified view of the neural net architecture. Layer sizes from left to right were actually 64, 32, 4, 8, 4, 1, as opposed to the 4, 3, 2, 3, 2, 1 shown here.</p>
</div>
<p>Evolution of model:</p>
<ol style="list-style-type: decimal">
<li>I started with only a couple of layers, with high dimensionality (around 64), predicting only a single protein expression. This was taking a long time to train the network with this large parameter space to train.</li>
<li>The training time went down a lot as I brought down the dimension of the layers, closer to 8 or 4. This didn’t seem to affect the performance, so I kept it to keep training times down.</li>
<li>I added a dropout layer, and this vastly improved performance.</li>
<li>I added the extra proteins, which actually decreased the performance of the couple proteins I was originally looking at.</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(keras)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"></a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="co"># set the input layer to be the right dimension</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4">inputs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/keras/man/layer_input.html">layer_input</a></span>(<span class="dt">shape =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(geneTrainingData)[<span class="dv">2</span>]))</a>
<a class="sourceLine" id="cb4-5" data-line-number="5"></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co"># create the first FOUR layers</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7">output0 &lt;-<span class="st"> </span>inputs <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="st">  </span><span class="co"># each of these lines creates a new layer, and specifies the number of nodes and the activation function</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/keras/man/layer_dropout.html">layer_dropout</a></span>(<span class="dt">rate =</span> <span class="dv">0</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/keras/man/layer_dense.html">layer_dense</a></span>(<span class="dt">units =</span> <span class="dv">256</span>, <span class="dt">activation =</span> <span class="st">'relu'</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/keras/man/layer_dense.html">layer_dense</a></span>(<span class="dt">units =</span> <span class="dv">128</span>, <span class="dt">activation =</span> <span class="st">'relu'</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/keras/man/layer_dense.html">layer_dense</a></span>(<span class="dt">units =</span> <span class="dv">64</span>, <span class="dt">activation =</span> <span class="st">'relu'</span>)</a>
<a class="sourceLine" id="cb4-13" data-line-number="13"></a>
<a class="sourceLine" id="cb4-14" data-line-number="14"><span class="co"># all of the layers are shared up until the last two, during which each protein gets its own set of 8 nodes leading to 1 output</span></a>
<a class="sourceLine" id="cb4-15" data-line-number="15">makeOutputLayer &lt;-<span class="st"> </span><span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb4-16" data-line-number="16">  <span class="kw"><a href="https://rdrr.io/pkg/keras/man/layer_dense.html">layer_dense</a></span>(output0, <span class="dt">units=</span><span class="dv">32</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="st">    </span><span class="kw"><a href="https://rdrr.io/pkg/keras/man/layer_dense.html">layer_dense</a></span>(<span class="dt">units=</span><span class="dv">1</span>, <span class="dt">name=</span> <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(proteinTrainingData)[x])</a>
<a class="sourceLine" id="cb4-18" data-line-number="18">}</a>
<a class="sourceLine" id="cb4-19" data-line-number="19"></a>
<a class="sourceLine" id="cb4-20" data-line-number="20"><span class="co"># for each protein, create the last 2 layers and add them to the original 4. See the architecture diagram</span></a>
<a class="sourceLine" id="cb4-21" data-line-number="21">outputs &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(proteinTrainingData)[<span class="dv">2</span>], makeOutputLayer)</a>
<a class="sourceLine" id="cb4-22" data-line-number="22"></a>
<a class="sourceLine" id="cb4-23" data-line-number="23">model &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/keras/man/keras_model.html">keras_model</a></span>(<span class="dt">inputs=</span>inputs, <span class="dt">outputs=</span>outputs)</a>
<a class="sourceLine" id="cb4-24" data-line-number="24"></a>
<a class="sourceLine" id="cb4-25" data-line-number="25">model <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/keras/man/reexports.html">compile</a></span>(</a>
<a class="sourceLine" id="cb4-26" data-line-number="26">  <span class="dt">loss =</span> <span class="st">"mse"</span>,</a>
<a class="sourceLine" id="cb4-27" data-line-number="27">  <span class="dt">optimizer =</span> <span class="kw"><a href="https://rdrr.io/pkg/keras/man/optimizer_rmsprop.html">optimizer_rmsprop</a></span>(<span class="dt">lr =</span> <span class="fl">.0001</span>), <span class="co"># learning rate here</span></a>
<a class="sourceLine" id="cb4-28" data-line-number="28">  <span class="dt">metrics =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="st">"mean_absolute_error"</span>),</a>
<a class="sourceLine" id="cb4-29" data-line-number="29">)</a>
<a class="sourceLine" id="cb4-30" data-line-number="30"></a>
<a class="sourceLine" id="cb4-31" data-line-number="31">model <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>()</a>
<a class="sourceLine" id="cb4-32" data-line-number="32"></a>
<a class="sourceLine" id="cb4-33" data-line-number="33"><span class="co"># Display training progress by printing a single dot for each completed epoch.</span></a>
<a class="sourceLine" id="cb4-34" data-line-number="34">print_dot_callback &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/keras/man/callback_lambda.html">callback_lambda</a></span>(</a>
<a class="sourceLine" id="cb4-35" data-line-number="35">  <span class="dt">on_epoch_end =</span> <span class="cf">function</span>(epoch, logs) {</a>
<a class="sourceLine" id="cb4-36" data-line-number="36">    <span class="cf">if</span> (epoch <span class="op">%%</span><span class="st"> </span><span class="dv">50</span> <span class="op">==</span><span class="st"> </span><span class="dv">0</span>) <span class="kw"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</a>
<a class="sourceLine" id="cb4-37" data-line-number="37">    <span class="kw"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="st">"."</span>)</a>
<a class="sourceLine" id="cb4-38" data-line-number="38">  }</a>
<a class="sourceLine" id="cb4-39" data-line-number="39">)    </a>
<a class="sourceLine" id="cb4-40" data-line-number="40"><span class="co"># epochs refers to how many passes through the data we make, and affects run time linearly. we can use early stopping techniques to set this, or we can just give it a set value to reach. 50 to 100 is usually good, but here I use 25 to help it train fast.</span></a>
<a class="sourceLine" id="cb4-41" data-line-number="41">epochs &lt;-<span class="st"> </span><span class="dv">50</span></a>
<a class="sourceLine" id="cb4-42" data-line-number="42">batch_size &lt;-<span class="st"> </span><span class="dv">128</span></a></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">42</span>)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="co"># Train the model</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3">history &lt;-<span class="st"> </span>model <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/keras/man/reexports.html">fit</a></span>(<span class="dt">x =</span> geneTrainingData, <span class="dt">y =</span> <span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(proteinTrainingData)[<span class="dv">2</span>], </a>
<a class="sourceLine" id="cb5-4" data-line-number="4">    <span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb5-5" data-line-number="5">        proteinTrainingData[, x]</a>
<a class="sourceLine" id="cb5-6" data-line-number="6">    }), <span class="dt">batch_size =</span> batch_size, <span class="dt">epochs =</span> epochs, <span class="dt">validation_split =</span> <span class="fl">0.2</span>, </a>
<a class="sourceLine" id="cb5-7" data-line-number="7">    <span class="dt">verbose =</span> <span class="dv">0</span>, <span class="dt">callbacks =</span> <span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(print_dot_callback))</a></code></pre></div>
<p>Here we can take a look at some interesting statistics from the model process.</p>
<p>First we see a set of plots that show the model’s loss over time for four proteins, CD45RA, CD8a, CD45RO, and CD279_PD_1. We can clearly see that as the model trains, it becomes more effective at predicting these values.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">lossPlots</span>()</a></code></pre></div>
<p><img src="DeepLearningWriteup_files/figure-html/lossPlots-1.png" width="700"><img src="DeepLearningWriteup_files/figure-html/lossPlots-2.png" width="700"></p>
<p>Here we can see a scatter plot showing on one axis the actual protein values for each cell, and on the other axis, the model’s predicted values for those same proteins.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw">scatters</span>()</a></code></pre></div>
<p><img src="DeepLearningWriteup_files/figure-html/scatters-1.png" width="700"><img src="DeepLearningWriteup_files/figure-html/scatters-2.png" width="700"><img src="DeepLearningWriteup_files/figure-html/scatters-3.png" width="700"><img src="DeepLearningWriteup_files/figure-html/scatters-4.png" width="700"><img src="DeepLearningWriteup_files/figure-html/scatters-5.png" width="700"></p>
<p>And finally, I printed out a simple list of the correlation between the model’s predictions and the actual values. As you can see, many of the proteins are not able to be predicted at all by the model.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw">correlations</span>()</a></code></pre></div>
<pre><code>## [1] "The correlation between predicted and actual values for CD3 is 0.903877581040914"
## [1] "The correlation between predicted and actual values for CD4 is 0.811989126513764"
## [1] "The correlation between predicted and actual values for CD8a is 0.627628193000274"
## [1] "The correlation between predicted and actual values for CD11b is 0.937457843624349"
## [1] "The correlation between predicted and actual values for CD14 is 0.937119894903895"
## [1] "The correlation between predicted and actual values for CD15 is 0.15436638969795"
## [1] "The correlation between predicted and actual values for CD16 is 0.715424356882438"
## [1] "The correlation between predicted and actual values for CD19 is 0.800357931837902"
## [1] "The correlation between predicted and actual values for CD20 is 0.828752658361578"
## [1] "The correlation between predicted and actual values for CD25 is 0.316042612612236"
## [1] "The correlation between predicted and actual values for CD27 is 0.836557695955479"
## [1] "The correlation between predicted and actual values for CD28 is 0.872294633108993"
## [1] "The correlation between predicted and actual values for CD34 is 0.225855527969671"
## [1] "The correlation between predicted and actual values for CD45RA is 0.788498848654816"
## [1] "The correlation between predicted and actual values for CD45RO is 0.741847881133926"
## [1] "The correlation between predicted and actual values for CD56 is 0.701139721387135"
## [1] "The correlation between predicted and actual values for CD62L is 0.526987969758485"
## [1] "The correlation between predicted and actual values for CD69 is 0.379576977744866"
## [1] "The correlation between predicted and actual values for CD80 is 0.0569178227599703"
## [1] "The correlation between predicted and actual values for CD86 is 0.883795728710354"
## [1] "The correlation between predicted and actual values for CD127 is 0.787138300629222"
## [1] "The correlation between predicted and actual values for CD137 is 0.0262545327160522"
## [1] "The correlation between predicted and actual values for CD197 is 0.331933111843249"
## [1] "The correlation between predicted and actual values for CD274 is 0.0717220233056332"
## [1] "The correlation between predicted and actual values for CD278 is 0.828248438755361"
## [1] "The correlation between predicted and actual values for CD335 is 0.55961777606049"
## [1] "The correlation between predicted and actual values for PD_1 is 0.429535712676894"
## [1] "The correlation between predicted and actual values for HLA_DR is 0.921049273966867"
## [1] "The correlation between predicted and actual values for TIGIT is 0.362535915994389"
## [1] "The sum of squared correlations is 12.8707715289298"
## [1] "The average correlation is 0.443819707894132"</code></pre>
</div>
<div id="discussion" class="section level1">
<h1 class="hasAnchor">
<a href="#discussion" class="anchor"></a>Discussion</h1>
<p>Overall the results are relatively good. Considering that we are predicting within a single cell type, a good fraction of the variance in the protein expression can be explained by the gene quantities. If the model contained multiple types of cells, a lot more of the variance probably could be explained, as it is easier to predict expression using cell type markers than using other genes.</p>
<p>This method can easily be applied to other data sets, as it is very extensible and makes almost no assumptions. In fact, it will likely have more impressive performance on other data sets, as it is able to lean on cell type markers.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#cut-up-the-data">Cut up the data</a></li>
      <li><a href="#cleaning-the-data">Cleaning the data</a></li>
      <li><a href="#select-data">Select Data</a></li>
      <li><a href="#building-the-model">Building the model</a></li>
      <li><a href="#discussion">Discussion</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Alex Nghiem.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
